ssl/statem/statem_lib.c

```diff
@@ -1007,6 +1007,28 @@ int ssl_choose_client_version(SSL *s, int version)
 }
 
 /*
+ * ssl_verify_client_session_version - Verify that provided SSL_SESSION matches
+ * the protocol negotiated with server.
+ *
+ * @s: client SSL handle.
+ *
+ * Returns 1 when it matches, 0 otherwise
+ */
+int ssl_verify_client_session_version(SSL *s)
+{
+  /* Empty session and no EAP-FAST */
+  if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {
```

Comment:
> Took some time to figure it out, but I think I got it. Put the session->ssl_version = assignment back,

Oh, I was thinking it should go slightly higher, near `s->session->session_id_length = session_id_len;`. This way you fix `ssl_version` in the fresh session case, but don't violate thread-safety in reuse.

> but not sure which check you think should be removed.

The one this thread is attached to. `if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {`. Or is that for something else? I think this function really just wants to be a version check.

> I don't really want to go into more involved approach, but again it may be a good idea to do this after this PR. Right now, I'm trying to fix a bug that node.js users are experiencing.

Sure. I did a minimal fix before the more aggressive one myself. :-)